### ISSUE TITLE
SP metadata not provided

### ISSUE THREAD

**Issue by dferch** [itzg/saml-auth-proxy#1]
I installed saml-auth-proxy from scratch with adfs backend. As described in your notes I want to get the sp metadata by running this command:

`curl http://localhost:19001/saml/metadata`

But I only got:

`404 page not found`

I‘m using the release 1.0.0 on RHEL x86. Any hint what happen ?

**Comment by itzg** [comment]
Just to confirm things, did you use `--bind` to change the bound listening port from 8080 to 19001? If not, something else at 19001 on your system would be answering that 404.

**Comment by itzg** [comment]
...and I just noticed my README is consistent about that port either. 8080 is definitely the default, not 8000 that I mention in another place there.

**Comment by dferch** [comment]
Yes, I changed the default port using `—bind` to 19001. The saml-auth-proxy is installed on a HAproxy cluster and I need more than one instance of the saml-auth-proxy. So the first instance will use 19001, second 19002, ... All implemented as systemd units grouped by a saml-auth-proxy.target. Seems to be the best way to run this little tool :-)

**Comment by itzg** [comment]
Yeah, we deploy the proxy behind an nginx ingress with the proxy as a side car to our app containers. With that said, it sounds HAproxy may not be routing the request as expected and is itself responding with the 404. 

I don’t have much debug logging in there yet, but I can work on adding some to help debug this.

**Comment by itzg** [comment]
One more thing to try is accessing the health endpoint to help me narrow down things. Try

    curl http://localhost:19001/_health

**Comment by dferch** [comment]
I already bypassed the HAproxy. I found the endpoints provided by the saml-auth-proxy in the source code and checked your code and the dependend go libraries (e.g. samlsp). If I request the URIs with curl, I‘ve got this results:

`
curl http://localhost:19001/_health
OK

curl http://localhost:19001/saml/metadata
404 page not found

curl http://localhost:19001/saml
<a href=„/saml/„>Moved Permanently</a>
` 

So it seems the proxy handle „/saml/„ correct, but the metadata is not provided by samlsp. I looked into crewjam/saml and someone mention, that in 0.2.0 the struct „Metadata“ has been renamed to „EntityDescriptor“. Does this matter ?

**Comment by dferch** [comment]
I found this at crewjam/saml:
[ACS 404-Not Foind #180](https://github.com/crewjam/saml/issues/180)

Looks like my issue :-|

**Comment by itzg** [comment]
Good find, but that issue may not be the same as yours since I get the impression that one is happening during IdP redirect.

In your case, I did find that the SAML middleware is building the path it expects from the path of the URL configured here https://github.com/crewjam/saml/blob/master/samlsp/samlsp.go#L40-L41

So...what are you providing for `--base-url`? If you're providing a path on that, then that would explain why it's not working and I can the `/saml` binding to be relative to the same path.

**Comment by dferch** [comment]
My base-url ist the url used to connect to the service on HAproxy. It‘s `https://grafana-test.my.domain.com:3000/`. And I tested the `.../saml/metadata` by using this base url. But it won‘t work too. I‘ll run some more tests tomorrow.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by itzg** [comment]
I finally got a chance to experiment some myself. I'm not sure yet why this makes a difference, but I found when I left off the trailing slash such as

    --base-url http://localhost.my.domain:3000

it worked. When I had a trailing slash like your usage I also got a 404:

    --base-url http://localhost.my.domain:3000/

I'm hopeful the same workaround will work for you, but I'll also investigate a proper solution. Speaking of which, I'm thinking of a PR on the crewjam module to allow for proxy URL re-write handling.
